core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/RuinMoveSelectorFactory.java

```diff
@@ -0,0 +1,50 @@
+package org.optaplanner.core.impl.heuristic.selector.move.generic;
+
+import java.util.*;
+
+import org.optaplanner.core.config.heuristic.selector.common.SelectionCacheType;
+import org.optaplanner.core.config.heuristic.selector.common.SelectionOrder;
+import org.optaplanner.core.config.heuristic.selector.entity.EntitySelectorConfig;
+import org.optaplanner.core.config.heuristic.selector.move.generic.RuinMoveSelectorConfig;
+import org.optaplanner.core.impl.domain.entity.descriptor.EntityDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
+import org.optaplanner.core.impl.domain.variable.descriptor.ShadowVariableDescriptor;
+import org.optaplanner.core.impl.heuristic.HeuristicConfigPolicy;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelector;
+import org.optaplanner.core.impl.heuristic.selector.entity.EntitySelectorFactory;
+import org.optaplanner.core.impl.heuristic.selector.move.AbstractMoveSelectorFactory;
+import org.optaplanner.core.impl.heuristic.selector.move.MoveSelector;
+
+public class RuinMoveSelectorFactory<Solution_>
+        extends AbstractMoveSelectorFactory<Solution_, RuinMoveSelectorConfig> {
+
+    public RuinMoveSelectorFactory(RuinMoveSelectorConfig moveSelectorConfig) {
+        super(moveSelectorConfig);
+    }
+
+    @Override
+    protected MoveSelector<Solution_> buildBaseMoveSelector(HeuristicConfigPolicy<Solution_> configPolicy,
+            SelectionCacheType minimumCacheType, boolean randomSelection) {
+        EntitySelectorConfig entitySelectorConfig =
+                Objects.requireNonNullElseGet(config.getEntitySelectorConfig(), EntitySelectorConfig::new);
+        EntitySelectorConfig secondaryEntitySelectorConfig =
+                Objects.requireNonNullElseGet(config.getSecondaryEntitySelectorConfig(), EntitySelectorConfig::new);
```

Comment:
I'm wondering... the two selectors exist for nearby only.
If the user decides not to use nearby, why do we have this complicated code handling two selectors? Why not introduce a much simpler variant with just the one selector, simply picking the entities to ruin?

Reply:
I separated these two, but I will refactor and better test them in an upcoming commit